Implement class ComputerState

yevGen-S opened the pull request:
**Added abstract class ComputerState**

- Added subclass NullComputerState 
- Added subclass PreviouslyUsedComputerState
- Added subclass OldComputerState
- Added subclass UsualComputerState
- Added subclass XiaomiMiNotebookState
- Added subclass MacbookState
- Properties have default values
   Closes: #4

Bazalii reviewed (commented):
1. Did you just merge Kirill's brunch into yours, to get his files or what? 
2. You also should implement all other children of ComputeState(Old, Previously used etc)
3. Please add to description of PR list of issues that will be closed by this request, after that they will link automatically.
4. You also have to add label for PR

yevGen-S commented:
>     1. Did you just merge Kirill's branch into yours, to get his files or what?
>Perhaps I accidentally merged Kiril's branch when I pulled up the latest updates of the project 
>     2. You also should implement all other children of ComputeState(Old, Previously used etc)
> fixed: i added children classes. And have some questions about it. For example, what about to add enumeration class for computerCourses. And also I think it's better to add owner:Player to ComputerState class.
>     3. Please add to description of PR list of issues that will be closed by this request, after that they will link automatically.
> fixed
>     4. You also have to add label for PR
fixed

Bazalii commented:
> > ```
> > 1. Did you just merge Kirill's branch into yours, to get his files or what?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Perhaps I accidentally merged Kiril's branch when I pulled up the latest updates of the project
> > 2. You also should implement all other children of ComputeState(Old, Previously used etc)
> > fixed: i added children classes. And have some questions about it. For example, what about to add enumeration class for computerCourses. And also I think it's better to add owner:Player to ComputerState class.
> > 3. Please add to description of PR list of issues that will be closed by this request, after that they will link automatically.
> > fixed
> > 4. You also have to add label for PR
> > fixed

We don't need the field owner at all because it was in the previous version of UML with mistakes and we added two inner classes: bag and courses, so there is no need in owner field. Please remove it from ComputerState. I askeArseny about enum. I also think that it is a good idea!

Bazalii reviewed (changes requested):
1. Where is available_courses initialization in child classes?
2. Where is override of the function isAvailable in child classes?
3. Check my other commentaries about your code
4. Could you rewrite the history to remove Kirill's files from this branch?

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/ComputerState.kt:
```diff
@@ -0,0 +1,21 @@
+
+    protected val available_courses: List<String> = listOf<String>()
+
+    abstract fun changeState(thingsHandler: Things)
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/NullComputerState.kt:
```diff
@@ -0,0 +1,16 @@
+package android.clicker.school_live_simulator
+
+open class NullComputerState: ComputerState {
+    override val price:Int = 0
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/ComputerState.kt:
```diff
@@ -0,0 +1,21 @@
+    abstract fun changeState(thingsHandler: Things)
+
+    /**
+     * This is a function to ckeck availability of the course
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/NullComputerState.kt:
```diff
@@ -0,0 +1,16 @@
+     *
+     * @param Player.bag player's bag of things
+     */
+
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/NullComputerState.kt:
```diff
@@ -0,0 +1,16 @@
+package android.clicker.school_live_simulator
+
+open class NullComputerState: ComputerState {
```
Fixed

yevGen-S commented:
2. Why should I override fun isAvailable in children if I we're waiting the same behaviour from it as in super class?

yevGen-S commented:
> 1. Where is available_courses initialization in child classes?
> 2. Where is override of the function isAvailable in child classes?
Why should I override fun isAvailable in children if I we're waiting the same behaviour from it as in super class?
> 3. Check my other commentaries about your code
> 4. Could you rewrite the history to remove Kirill's files from this branch?

yevGen-S commented:
> > 1. Where is available_courses initialization in child classes?
Fixed
> > 2. Where is override of the function isAvailable in child classes?
> >    Why should I override fun isAvailable in children if I we're waiting the same behaviour from it as in super class?
> > 3. Check my other commentaries about your code
Fixed
> > 4. Could you rewrite the history to remove Kirill's files from this branch?
Fixed

Bazalii commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/MacbookState.kt:
```diff
@@ -0,0 +1,13 @@
+package android.clicker.school_live_simulator
+
+open class MacbookState: XiaomiMiNotebookState() {
```
You have to add one more space here in every child class
```suggestion
open class MacbookState : XiaomiMiNotebookState() {
```

yevGen-S replied:
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/ComputerState.kt:
```diff
@@ -0,0 +1,35 @@
+     * @return True if the course is available
+     */
+
+    open fun isAvailable(courseName: KClass<*>): Boolean {
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/PreviouslyUsedComputerState.kt:
```diff
@@ -0,0 +1,14 @@
+
+    init {
+
+    }
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/OldComputerState.kt:
```diff
@@ -0,0 +1,14 @@
+package android.clicker.school_live_simulator
+
+open class OldComputerState: PreviouslyUsedComputerState() {
+    override  price: Int = 10000
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/UsualComputerState.kt:
```diff
@@ -0,0 +1,14 @@
+package android.clicker.school_live_simulator
+
+open class UsualComputerState: OldComputerState() {
+    override  price: Int = 30000
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/MacbookState.kt:
```diff
@@ -0,0 +1,13 @@
+package android.clicker.school_live_simulator
+
+open class MacbookState: XiaomiMiNotebookState() {
+    override  price: Int = 200000
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/XiaomiMiNotebookState.kt:
```diff
@@ -0,0 +1,14 @@
+package android.clicker.school_live_simulator
+
+open class XiaomiMiNotebookState: UsualComputerState() {
+    override  price: Int = 60000
```
Fixed

yevGen-S commented on Project_source/app/src/main/java/android/clicker/school_live_simulator/ComputerState.kt:
```diff
@@ -0,0 +1,35 @@
+     * Computer price
+     */
+
+    protected abstract val price: Int
```
Fixed: I made this field public and there is no need in private set ,because it's val

Bazalii reviewed (approved):
Everything looks great, waiting for @Ferrociraptus commentaries

Ferrociraptus reviewed (approved):
Good job!
